lib/semian/circuit_breaker.rb

```diff
@@ -43,70 +44,77 @@ def mark_failed(_error)
 
     def mark_success
       return unless half_open?
-      @successes += 1
+      @successes.increment
       close if success_threshold_reached?
     end
 
     def reset
-      @errors = []
-      @successes = 0
+      @errors.clear
+      @successes.value = 0
       close
     end
 
+    def destroy
+      @errors.destroy
+      @successes.destroy
+      @state.destroy
+    end
+
     private
 
     def closed?
-      state == :closed
+      @state.value == :closed
     end
 
     def close
       log_state_transition(:closed)
-      @state = :closed
-      @errors = []
+      @state.value = :closed
+      @errors.clear
     end
 
     def open?
-      state == :open
+      @state.value == :open
     end
 
     def open
       log_state_transition(:open)
-      @state = :open
+      @state.value = :open
     end
 
     def half_open?
-      state == :half_open
+      @state.value == :half_open
     end
 
     def half_open
       log_state_transition(:half_open)
-      @state = :half_open
-      @successes = 0
+      @state.value = :half_open
+      @successes.value = 0
     end
 
     def success_threshold_reached?
-      @successes >= @success_count_threshold
+      @successes.value >= @success_count_threshold
     end
 
     def error_threshold_reached?
-      @errors.count == @error_count_threshold
+      @errors.size == @error_count_threshold
     end
 
     def error_timeout_expired?
-      @errors.last && (@errors.last + @error_timeout < Time.now)
+      time_ms = @errors.last
+      time_ms && (Time.at(time_ms / 1000) + @error_timeout < Time.now)
     end
 
-    def push_time(window, max_size, duration:, time: Time.now)
-      window.shift while window.first && window.first + duration < time
-      window.shift if window.size == max_size
-      window << time
+    def push_time(window, duration:, time: Time.now)
+      # The sliding window stores the integer amount of milliseconds since epoch as a timestamp
+      window.shift while window.first && Time.at(window.first / 1000) + duration < time
```

Comment:
Do we actually need to cast to `Time` here? A pure integer timestamp comparison wouldn't do it?